### DISCUSSION TITLE
`list[UploadFile] | None` only works with Pydantic 1

### DISCUSSION THREAD

**Discussion by pbasista** [fastapi/fastapi#11131] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, UploadFile
import uvicorn

app = FastAPI()


@app.post("/upload")
def handle_multiple_file_upload(
    file_field: list[UploadFile] | None,
):
    return str(file_field)


if __name__ == "__main__":
    uvicorn.run(app)
```

### Description

* Install a recent version of FastAPI which by default installs Pydantic 2 as a dependency.
* The provided script creates an endpoint which responds with the value of the `file_field` parameter. Run it (i.e. `python script.py`).
* Create files `file1.txt` and `file2.txt` with arbitrary content.
* Run `curl 'http://localhost:8000/upload' -F file_field=@file1.txt -F file_field=@file2.txt`.
* The FastAPI server responds with HTTP status 422 Unprocessable Entity. The response body contains the error details:
```json
{"detail":[{"type":"list_type","loc":["body","file_field"],"msg":"Input should be a valid list","input":{"filename":"file2.txt","file":{"_file":{},"_max_size":1048576,"_rolled":false,"_TemporaryFileArgs":{"mode":"w+b","buffering":-1,"suffix":null,"prefix":null,"encoding":null,"newline":null,"dir":null,"errors":null}},"size":20,"headers":{"content-disposition":"form-data; name=\"file_field\"; filename=\"file2.txt\"","content-type":"text/plain"}},"url":"https://errors.pydantic.dev/2.6/v/list_type"}]}
```

It appears as if only the second file has been provided by FastAPI as the value to the `file_field` parameter.

If the type annotation of `file_field` is changed to be non-optional, i.e. from `list[UploadFile] | None` to `list[UploadFile]`, then the endpoint would respond with a correct HTTP response status 200. The `file_field` would contain both uploaded files:

```
"[UploadFile(filename='file1.txt', size=19, headers=Headers({'content-disposition': 'form-data; name=\"file_field\"; filename=\"file1.txt\"', 'content-type': 'text/plain'})), UploadFile(filename='file2.txt', size=20, headers=Headers({'content-disposition': 'form-data; name=\"file_field\"; filename=\"file2.txt\"', 'content-type': 'text/plain'}))]"
```

This issue cannot be reproduced with Pydantic 1 where the type annotation `list[UploadFile] | None` seems to be handled correctly. To reproduce:

* Undo the change to the type annotation of `file_field` and keep it optional, i.e. `list[UploadFile] | None`.
* Install Pydantic 1 while keeping everything else the same. This can be achieved e.g. by running `pip install 'pydantic<2'`. At the moment this command results in Pydantic version `1.10.14` being installed.
* Restart the provided Python script (i.e. `python script.py`).
* Run `curl 'http://localhost:8000/upload' -F file_field=@file1.txt -F file_field=@file2.txt`.
* Now the endpoint responds with HTTP status 200 and the `file_field` parameter contains the expected *list* of uploaded files, even when its type annotation is `list[UploadFile] | None`:

This issue is definitely related to Pydantic. But it seems to me that it is not *caused* by Pydantic but rather by FastAPI passing incorrect values to the route handler parameters in certain situations. I am, however, unsure how to properly debug this.

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

Python 3.11.3

### Additional Context

Current documentation related to uploading multiple files:

* https://fastapi.tiangolo.com/tutorial/request-files/#multiple-file-uploads

Loosely related questions or issues:

* https://github.com/tiangolo/fastapi/discussions/10987
* https://github.com/tiangolo/fastapi/issues/10286

**Comment by YuriiMotov** [comment]
This is a bug of FastAPI and [this PR](https://github.com/tiangolo/fastapi/pull/9928) fixes it (but it hasn't merged yet).
BTW, to make this field displayed correctly in openapi schema, define it as `file_field: list[UploadFile] | SkipJsonSchema[None]` (this is another one bug)

**Comment by RomanChumakov** [comment, marked as answer]
The bug was still relevant at version 0.111.0. I updated to 0.113.0 and everything worked fine
